src/hato/client.clj

```diff
@@ -284,42 +275,114 @@
+      (.handle
+        (reify BiFunction
+          (apply [_ resp e]
+            [resp (unwrap-async-exception e)])))
+      (.thenCompose
+        (reify Function
+          (apply [_ [resp e]]
+            (let [retry-fut (if retry-handler
+                              (retry-handler resp e req retry-count)
+                              (CompletableFuture/completedFuture false))]
+              (-> (if (instance? CompletableFuture retry-fut)
+                    ^CompletableFuture retry-fut
+                    (CompletableFuture/completedFuture retry-fut))
+                  (.thenCompose
+                    (reify Function
+                      (apply [_ retry?]
+                        (if retry?
+                          (request-with-retries* http-client http-request body-handler req retry-handler (inc retry-count))
+                          (if e
+                            (CompletableFuture/failedFuture e)
+                            (CompletableFuture/completedFuture
+                              (response-map {:request     req
+                                             :http-client http-client
+                                             :response    resp}))))))))))))))
+
 (defn request*
-  [{:keys [http-client async? as]
+  [{:keys [http-client async? retry-handler]
     :as   req} & [respond raise]]
   (let [^HttpClient http-client (if (instance? HttpClient http-client) http-client (build-http-client http-client))
-        http-request (ring-request->HttpRequest req)
-        bh (->BodyHandler as)]
-    (if-not async?
-      (let [resp (.send http-client http-request bh)]
-        (response-map
-         {:request     req
-          :http-client http-client
-          :response    resp}))
-
-      (-> (.sendAsync http-client http-request bh)
+        http-request            (ring-request->HttpRequest req)
+        retry-handler           (if (= :auto retry-handler) default-retry-handler retry-handler)
+        body-handler            (HttpResponse$BodyHandlers/ofInputStream)
+        resp-fut                (request-with-retries* http-client http-request body-handler req retry-handler 0)]
+    (if async?
+      (-> resp-fut
           (.thenApply
-           (reify Function
-             (apply [_ resp]
-               (respond
-                (response-map
-                 {:request     req
-                  :http-client http-client
-                  :response    resp})))))
+            (reify Function
+              (apply [_ resp-map]
+                (respond resp-map))))
           (.exceptionally
-           (reify Function
-             (apply [_ e]
-               (let [cause (.getCause ^Exception e)]
-                 (if (instance? ExceptionInfo cause)
-                   (raise cause)
-                   (raise e))))))))))
+            (reify Function
+              (apply [_ e]
+                (raise (unwrap-async-exception e))))))
+      (try @resp-fut
```

Comment:
Since `.send` is still async (it just does a `.get` internally), I don't think we're losing anything by handling it this way other than we must unwrap ExecutionException